### PR TITLE
colby::conversions.hpp: Roll our own Lab conversions

### LINKED ISSUE #10: RGB2LAB Colorspace Conversion Issues
There are some discrepancies between the SP3000 algorithm color space converters and the ones that we utilise via OpenCV. 

SP3000 color conversion formulas come from [this page](http://www.easyrgb.com/?X=MATH), and seem to assume sRGB.  It also scales the values to be within the 0-100 range. 

The [OpenCV implementation of RGB2Lab](http://docs.opencv.org/2.4/modules/imgproc/doc/miscellaneous_transformations.html) does not perform the gamma correction; nor does it scale the values to be in the 0-100 scale.

It might be worthwhile to roll our own RGB2Lab converter so that we can more finely control the conversion and then use the same color sensitivity parameters as the SP3000 algorithm.

### PR DESCRIPTION
**Pull request by jordan-heemskerk**
- Write own conversions based on SP3000 routines
 - Update parameters to match SP3000 tolerances
 - Replace OpenCV conversions throughout with our own
 - Tests for our own implementation, using values generated from SP3000
   routines

Fixes #10

### CODE REVIEW

File: include/colby/conversions.hpp [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+	for (int i = 0; i < bgr.cols; ++i) {
+		for (int j = 0; j < bgr.rows; ++j) {
+			retr.at<cv::Vec3f>(j,i) = bgr2lab(bgr.at<cv::Vec3b>(j,i));
+		}
+	}
+	return retr;
+}
+
+/**
+ *	converts a cv::Mat of 32-bit floating point CIE Lab values
```
Comment (RobertLeahy) [inline include/colby/conversions.hpp]: Capitalization

I've seen "CIELAB" on [Wikipedia](https://en.wikipedia.org/wiki/Lab_color_space#CIELAB)

File: include/colby/conversions.hpp [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+ *		The color in BGR
+ *	\returns
+ *		The color in CIE Lab space
+ */
+inline cv::Vec3f bgr2lab (const cv::Vec3b bgr) noexcept {
+	return xyz2lab(bgr2xyz(bgr));
+}
+
+/**
+ *	converts a 32-bit floating point CIE Lab value
```
Comment (RobertLeahy) [inline include/colby/conversions.hpp]: Capitalization

I've seen "CIELAB" on [Wikipedia](https://en.wikipedia.org/wiki/Lab_color_space#CIELAB)

File: include/colby/conversions.hpp [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+	if (r > 255.f) r = 255.f;
+	if (g > 255.f) g = 255.f;
+	if (b > 255.f) b = 255.f;
+
+	return cv::Vec3b(b,g,r);
+}
+
+/**
+ *	converts a 8-bit bgr value to 32-bit floating
```
Comment (RobertLeahy) [inline include/colby/conversions.hpp]: Capitalization: "converts" => "Converts" & "bgr" => "BGR"

File: include/colby/conversions.hpp [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+	if (r < 0.f) r = 0.f;
+	if (g < 0.f) g = 0.f;
+	if (b < 0.f) b = 0.f;
+
+	if (r > 255.f) r = 255.f;
+	if (g > 255.f) g = 255.f;
+	if (b > 255.f) b = 255.f;
+
+	return cv::Vec3b(b,g,r);
+}
```
Comment (RobertLeahy) [inline include/colby/conversions.hpp]: Are all of these functions implementation details, is that why they're undocumented?

File: include/colby/conversions.hpp [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+ *	\param [in] bgr
+ *		A cv::Mat of BGR values
+ *	\returns
+ *		A cv::Mat of Lab values
+ */
+inline cv::Mat bgr2lab(const cv::Mat & bgr) {
+	if (bgr.type() != CV_8UC3) throw std::logic_error("Expected 3 channel 8 bit image");
+	cv::Mat retr(bgr.rows, bgr.cols, CV_32FC3);
+	for (int i = 0; i < bgr.cols; ++i) {
+		for (int j = 0; j < bgr.rows; ++j) {
```
Comment (RobertLeahy) [inline include/colby/conversions.hpp]: Use [`std::transform`](http://en.cppreference.com/w/cpp/algorithm/transform) with raw pointers `reinterpret_cast` from `cv::Mat::data`, this is not only more readable (due to the use of a standard algorithm) but should also be marginally more efficient

File: include/colby/conversions.hpp [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+ *	\param [in] lab
+ *		A cv::Mat of colors in CIE Lab space
+ *	\returns
+ *		A cv::Mat of colors in BGR space
+ */
+inline cv::Mat lab2bgr(const cv::Mat & lab) {
+	if (lab.type() != CV_32FC3) throw std::logic_error("Expected 3 channel 32 bit floating point image");
+	cv::Mat retr(lab.rows, lab.cols, CV_8UC3);
+	for (int i = 0; i < lab.cols; ++i) {
+		for (int j = 0; j < lab.rows; ++j) {
```
Comment (RobertLeahy) [inline include/colby/conversions.hpp]: Use [`std::transform`](http://en.cppreference.com/w/cpp/algorithm/transform) with raw pointers `reinterpret_cast` from `cv::Mat::data`, this is not only more readable (due to the use of a standard algorithm) but should also be marginally more efficient

File: src/colby/sp3000_color_by_numbers.cpp [hunk 7]
```diff
@@ -172,20 +173,6 @@ cv::Mat sp3000_color_by_numbers::graph::mat () const {
...
-	return retr;
-}
-
-cv::Mat sp3000_color_by_numbers::convert_lab_to_bgr (const cv::Mat & img) {
-	cv::Mat retr;
-	cv::cvtColor(img,retr,CV_RGB2BGR);
-	if (!retr.data) throw std::runtime_error("cv::cvtColor failed to convert image to BGR");
-	return retr;
-}
-
```
Comment (RobertLeahy) [inline src/colby/sp3000_color_by_numbers.cpp]: Remove forward declarations from .hpp file

File: src/colby/test/conversions.cpp [hunk 8]
```diff
@@ -0,0 +1,148 @@
+#include <colby/conversions.hpp>
+#include <opencv2/core/matx.hpp>
+#include <cmath>
+#include <limits>
```
Comment (RobertLeahy) [inline src/colby/test/conversions.cpp]: Why is this header needed?

File: include/colby/conversions.hpp [hunk 9]
```diff
@@ -0,0 +1,164 @@
+/**
+ *	\file
+ */
+
+#pragma once
+
+#include <opencv2/core/mat.hpp>
+#include <opencv2/core/matx.hpp>
+#include <iostream>
```
Comment (RobertLeahy) [inline include/colby/conversions.hpp]: Why is this needed?

File: include/colby/conversions.hpp [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+
+	auto y = (l+16.f)/116.f;
+	auto x = a/500.f+y;
+	auto z = y-b/200.f;
+
+	x = std::pow(x, 3.f)>0.008856 ? std::pow(x, 3.f)  : (x-16.f/116.f)/7.787f;
+	y = std::pow(y, 3.f)>0.008856 ? std::pow(y, 3.f)  : (y-16.f/116.f)/7.787f;
+	z = std::pow(z, 3.f)>0.008856 ? std::pow(z, 3.f)  : (z-16.f/116.f)/7.787f;
+
+	x *= 95.047;
```
Comment (RobertLeahy) [inline include/colby/conversions.hpp]: This is a double literal

File: include/colby/conversions.hpp [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+	auto a = lab[1];
+	auto b = lab[2];
+
+	auto y = (l+16.f)/116.f;
+	auto x = a/500.f+y;
+	auto z = y-b/200.f;
+
+	x = std::pow(x, 3.f)>0.008856 ? std::pow(x, 3.f)  : (x-16.f/116.f)/7.787f;
+	y = std::pow(y, 3.f)>0.008856 ? std::pow(y, 3.f)  : (y-16.f/116.f)/7.787f;
+	z = std::pow(z, 3.f)>0.008856 ? std::pow(z, 3.f)  : (z-16.f/116.f)/7.787f;
```
Comment (RobertLeahy) [inline include/colby/conversions.hpp]: 0.008856 is a double literal which will cause promotion before the comparison

File: include/colby/conversions.hpp [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+	g = g > 0.0031308f ? 1.055f*std::pow(g, 1.f/2.4f)-0.055f : 12.92f*g;
+	b = b > 0.0031308f ? 1.055f*std::pow(b, 1.f/2.4f)-0.055f : 12.92f*b;
+
+	r *= 255.f;
+	g *= 255.f;
+	b *= 255.f;
+
+	if (r < 0.f) r = 0.f;
+	if (g < 0.f) g = 0.f;
+	if (b < 0.f) b = 0.f;
```
Comment (RobertLeahy) [inline include/colby/conversions.hpp]: You can just use 0, 0 has special rules that allow it to be a pointer, integer, and floating point constant

File: src/colby/sp3000_color_by_numbers.cpp [hunk 13]
```diff
@@ -397,11 +384,11 @@ sp3000_color_by_numbers::result sp3000_color_by_numbers::convert_impl (const cv:
 		immediate_image_factory () = delete;
 		immediate_image_factory (const cv::Mat & mat) noexcept : mat_(mat) {	}
 		virtual cv::Mat image () override {
-			return convert_lab_to_bgr(mat_);
+			return lab2bgr(mat_);
 		}
 	};
 	//	1. Convert the pixels to the CIELAB colour space
-	auto lab = convert_bgr_to_lab(src);
+	cv::Mat lab = bgr2lab(src);
```
Comment (RobertLeahy) [inline src/colby/sp3000_color_by_numbers.cpp]: The fact that `auto` doesn't work (or doesn't seem to work) really bothers me especially since I strongly prefer [AAA style](https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/)

Let me take a look at this later...

File: src/colby/test/conversions.cpp [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+
+	GIVEN("A cv::Mat of Lab values") {
+		cv::Mat lab(cv::Mat::zeros(2,3,CV_32FC3));
+		lab.at<cv::Vec3f>(0,2) = cv::Vec3f(34.54f,10.8f,-3.2);
+		lab.at<cv::Vec3f>(1,0) = cv::Vec3f(100.f,0.f,0.f);
+		WHEN("It is converted to Lab") {
+			cv::Mat bgr = lab2bgr(lab);
+			THEN("Each value of the cv::Mat is converted correctly") {
+
+				auto black = bgr.at<cv::Vec3b>(0,0);
```
Comment (RobertLeahy) [inline src/colby/test/conversions.cpp]: `auto &&`

File: src/colby/test/conversions.cpp [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+				CHECK(std::abs(int(black[2])) < 2);
+				CHECK(std::abs(int(black[1])) < 2);
+				CHECK(std::abs(int(black[0])) < 2);
+
+				auto arbitrary = bgr.at<cv::Vec3b>(0,2);
```
Comment (RobertLeahy) [inline src/colby/test/conversions.cpp]: `auto &&`

File: src/colby/test/conversions.cpp [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+				CHECK(std::abs(int(arbitrary[2]) - 96) < 2);
+				CHECK(std::abs(int(arbitrary[1]) - 75) < 2);
+				CHECK(std::abs(int(arbitrary[0]) - 87) < 2);
+
+				auto white = bgr.at<cv::Vec3b>(1,0);
```
Comment (RobertLeahy) [inline src/colby/test/conversions.cpp]: `auto &&`

**Review by RobertLeahy — CHANGES_REQUESTED** [review]

File: include/colby/conversions.hpp (hunk 9 above)
Comment (jordan-heemskerk) [inline include/colby/conversions.hpp]: removed

File: include/colby/conversions.hpp (hunk 11 above)
Comment (jordan-heemskerk) [inline include/colby/conversions.hpp]: fixed

File: include/colby/conversions.hpp (hunk 10 above)
Comment (jordan-heemskerk) [inline include/colby/conversions.hpp]: fixed

File: include/colby/conversions.hpp (hunk 12 above)
Comment (jordan-heemskerk) [inline include/colby/conversions.hpp]: fixed

File: include/colby/conversions.hpp (hunk 4 above)
Comment (jordan-heemskerk) [inline include/colby/conversions.hpp]: No I think I was just lazy, it's entirely possible for each to be used independently. I will document them all I guess.

File: include/colby/conversions.hpp (hunk 3 above)
Comment (jordan-heemskerk) [inline include/colby/conversions.hpp]: done

File: include/colby/conversions.hpp (hunk 2 above)
Comment (jordan-heemskerk) [inline include/colby/conversions.hpp]: done

File: include/colby/conversions.hpp (hunk 5 above)
Comment (jordan-heemskerk) [inline include/colby/conversions.hpp]: Can't seem to figure this out on my own...

File: include/colby/conversions.hpp (hunk 1 above)
Comment (jordan-heemskerk) [inline include/colby/conversions.hpp]: done

File: src/colby/sp3000_color_by_numbers.cpp (hunk 7 above)
Comment (jordan-heemskerk) [inline src/colby/sp3000_color_by_numbers.cpp]: done

File: src/colby/sp3000_color_by_numbers.cpp (hunk 13 above)
Comment (jordan-heemskerk) [inline src/colby/sp3000_color_by_numbers.cpp]: sounds good

File: src/colby/test/conversions.cpp (hunk 8 above)
Comment (jordan-heemskerk) [inline src/colby/test/conversions.cpp]: the trick is, it isn't

File: src/colby/test/conversions.cpp (hunk 14 above)
Comment (jordan-heemskerk) [inline src/colby/test/conversions.cpp]: done

File: src/colby/test/conversions.cpp (hunk 15 above)
Comment (jordan-heemskerk) [inline src/colby/test/conversions.cpp]: done

File: src/colby/test/conversions.cpp (hunk 16 above)
Comment (jordan-heemskerk) [inline src/colby/test/conversions.cpp]: done

**Review by RobertLeahy — APPROVED** [review]
